system/lib/wasm_worker/library_wasm_worker.c

```diff
@@ -21,8 +22,18 @@ void __wasm_init_tls(void *memory);
 __attribute__((constructor(48)))
 static void emscripten_wasm_worker_main_thread_initialize() {
 	uintptr_t* sbrk_ptr = emscripten_get_sbrk_ptr();
-	__wasm_init_tls((void*)*sbrk_ptr);
-	*sbrk_ptr += __builtin_wasm_tls_size();
+	uintptr_t sbrk_addr = *sbrk_ptr;
+
+	if(__builtin_wasm_tls_align() != 0)
+	{
+		sbrk_addr = (sbrk_addr + __builtin_wasm_tls_align() - 1) & (-__builtin_wasm_tls_align());
+	}
```

Comment:
While the above rounding is technically correct, it would be very surprising if `__builtin_wasm_tls_align()` would ever be larger than 16 bytes, which is already the sbrk alignment. (If it was, there are other places in code we will need to audit) So let's keep it simple and go with an assert() route here instead, i.e.

```cpp
static void emscripten_wasm_worker_main_thread_initialize() {
  uintptr_t* sbrk_ptr = emscripten_get_sbrk_ptr();
  assert(*sbrk_ptr & 15 == 0);
  assert(__builtin_wasm_tls_align() <= 16);
  __wasm_init_tls((void*)*sbrk_ptr);
  *sbrk_ptr += (__builtin_wasm_tls_size() + 15) & -16;
}
```
There is a discrepancy between malloc alignment and sbrk alignment. Sbrk alignment is always fixed 16 bytes, so better not to refer to max_align_t here.

Reply:
@juj Cool, that's what I had in the initial version(16), to match emscripten_wasm_worker_initialize. Made the changes.

Reply:
Yeh, looks great now, lean and to the point.

Btw, would you happen to have a small test at hand that would exercise this issue? It would be great to have a unit test that validates this scenario that the implementation was missing before?

Reply:
Sure, I will try to make a test case for this, based on the sample in the linked issue.